File: FBSimulatorControl/Strategies/FBSimulatorBootVerificationStrategy.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorBootVerificationStrategy.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimDeviceBootInfo.h>

#import <FBControlCore/FBControlCore.h>

#import "FBSimulator.h"
#import "FBSimulatorError.h"

@interface FBSimulatorBootVerificationStrategy ()

@property (nonatomic, strong, readonly) FBSimulator *simulator;
@property (nonatomic, strong, nullable, readwrite) SimDeviceBootInfo *lastBootInfo;
@property (nonatomic, strong, nullable, readwrite) NSDate *lastInfoUpdateDate;

@end

@implementation FBSimulatorBootVerificationStrategy

#pragma mark Initializers

+ (FBFuture<NSNull *> *)verifySimulatorIsBooted:(FBSimulator *)simulator
{
  return [[[FBSimulatorBootVerificationStrategy alloc] initWithSimulator:simulator] verifySimulatorIsBooted];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

#pragma mark Public

static NSTimeInterval BootVerificationWaitInterval = 0.5; // 500ms
static NSTimeInterval BootVerificationStallInterval = 1.5; // 60s

- (FBFuture<NSNull *> *)verifySimulatorIsBooted
{
  FBSimulator *simulator = self.simulator;

  return [[simulator
    resolveState:FBiOSTargetStateBooted]
    onQueue:simulator.workQueue fmap:^FBFuture *(NSNull *_) {
      return [FBFuture onQueue:simulator.workQueue resolveUntil:^{
        return [[self performBootVerification] delay:BootVerificationWaitInterval];
      }];
    }];
}

#pragma mark Private

- (FBFuture<NSNull *> *)performBootVerification
{
  SimDeviceBootInfo *bootInfo = self.simulator.device.bootStatus;
  if (!bootInfo) {
    return [[FBSimulatorError
      describeFormat:@"No bootInfo for %@", self.simulator]
      failFuture];
  }
  [self updateBootInfo:bootInfo];
  if (bootInfo.isTerminalStatus == NO) {
    return [[FBSimulatorError
      describeFormat:@"Not terminal status, status is %@", bootInfo]
      failFuture];
  }
  return FBFuture.empty;
}

- (void)updateBootInfo:(SimDeviceBootInfo *)bootInfo
{
  // The isEqual Method implementation does *not* take into account -[SimDeviceBootInfo bootElapsedTime].
  // We can check that the differences between the last info and the current one are greater some 'stall threshold.
  // This can be indicative that something has gone wrong in the boot process.
  NSTimeInterval stallInterval = BootVerificationStallInterval;
  id<FBControlCoreLogger> logger = self.simulator.logger;
  if (!self.lastInfoUpdateDate) {
    self.lastInfoUpdateDate = NSDate.date;
  }
  if ([bootInfo isEqual:self.lastBootInfo]) {
    NSTimeInterval updateInterval = [NSDate.date timeIntervalSinceDate:self.lastInfoUpdateDate];
    if (updateInterval < stallInterval) {
      return;
    }
    [logger logFormat:@"Boot Status has not changed from '%@' for %f seconds", [FBSimulatorBootVerificationStrategy describeBootInfo:bootInfo], updateInterval];
  } else {
    [logger.debug logFormat:@"Boot Status Changed: %@", [FBSimulatorBootVerificationStrategy describeBootInfo:bootInfo]];
    self.lastBootInfo = bootInfo;
    self.lastInfoUpdateDate = NSDate.date;
  }
}

+ (NSString *)describeBootInfo:(SimDeviceBootInfo *)bootInfo
{
  NSString *regular = [self regularBootInfo:bootInfo];
  NSString *migration = [self dataMigrationString:bootInfo];
  if (!migration) {
    return regular;
  }
  return [NSString stringWithFormat:@"%@ | %@", regular, migration];
}

+ (NSString *)regularBootInfo:(SimDeviceBootInfo *)bootInfo
{
  return [NSString stringWithFormat:
    @"%@ | Elapsed %f",
    [self bootStatusString:bootInfo.status],
    bootInfo.bootElapsedTime
  ];
}

+ (NSString *)bootStatusString:(SimDeviceBootInfoStatus)status
{
  switch (status) {
    case SimDeviceBootInfoStatusBooting:
      return @"Booting";
    case SimDeviceBootInfoStatusWaitingOnBackboard:
      return @"WaitingOnBackboard";
    case SimDeviceBootInfoStatusWaitingOnDataMigration:
      return @"WaitingOnDataMigration";
    case SimDeviceBootInfoStatusWaitingOnSystemApp:
      return @"WaitingOnSystemApp";
    case SimDeviceBootInfoStatusFinished:
      return @"Finished";
    case SimDeviceBootInfoStatusDataMigrationFailed:
      return @"DataMigrationFailed";
    default:
      return @"Unknown";
  }
}

+ (NSString *)dataMigrationString:(SimDeviceBootInfo *)bootInfo
{
  if (bootInfo.status != SimDeviceBootInfoStatusWaitingOnDataMigration) {
    return nil;
  }
  return [NSString stringWithFormat:
    @"Migration Phase '%@' | Migration Elapsed %f",
    bootInfo.migrationPhaseDescription,
    bootInfo.migrationElapsedTime
  ];
}

@end

File: FBSimulatorControl/Strategies/FBSimulatorBootVerificationStrategy.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

public enum FBSimulatorBootVerificationError: Error {
  case noBootInfo(simulatorDescription: String)
  case notTerminalStatus(bootInfoDescription: String)
}

extension FBSimulatorBootVerificationError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .noBootInfo(simulatorDescription):
      return "No bootInfo for \(simulatorDescription)"
    case let .notTerminalStatus(bootInfoDescription):
      return "Not terminal status, status is \(bootInfoDescription)"
    }
  }
}

public final class FBSimulatorBootVerificationStrategy {

  private let simulator: FBSimulator
  private var lastBootInfo: SimDeviceBootInfo?
  private var lastInfoUpdateDate: Date?

  // MARK: - Constants

  private static let bootVerificationWaitInterval: TimeInterval = 0.5
  private static let bootVerificationStallInterval: TimeInterval = 1.5

  private init(simulator: FBSimulator) {
    self.simulator = simulator
  }

  static func verifySimulatorIsBooted(_ simulator: FBSimulator) async throws {
    try await FBiOSTargetResolveState(simulator, .booted)
    let strategy = FBSimulatorBootVerificationStrategy(simulator: simulator)
    let interval = UInt64(bootVerificationWaitInterval * Double(NSEC_PER_SEC))
    while true {
      try Task.checkCancellation()
      try await Task.sleep(nanoseconds: interval)
      do {
        try strategy.performBootVerificationCheck()
        return
      } catch {
        // continue polling
      }
    }
  }

  private func performBootVerificationCheck() throws {
    let bootInfo: SimDeviceBootInfo? = simulator.device.bootStatus()
    guard let bootInfo else {
      throw FBSimulatorBootVerificationError.noBootInfo(simulatorDescription: String(describing: simulator))
    }
    updateBootInfo(bootInfo)
    if bootInfo.isTerminalStatus == false {
      throw FBSimulatorBootVerificationError.notTerminalStatus(bootInfoDescription: String(describing: bootInfo))
    }
  }

  private func updateBootInfo(_ bootInfo: SimDeviceBootInfo) {
    let stallInterval = FBSimulatorBootVerificationStrategy.bootVerificationStallInterval
    let logger = simulator.logger

    let updateDate = lastInfoUpdateDate ?? Date()
    if lastInfoUpdateDate == nil {
      lastInfoUpdateDate = updateDate
    }
    if bootInfo.isEqual(lastBootInfo) {
      let updateInterval = Date().timeIntervalSince(updateDate)
      if updateInterval < stallInterval {
        return
      }
      logger.log("Boot Status has not changed from '\(bootInfo.progressDescription)' for \(updateInterval) seconds")
    } else {
      logger.debug().log("Boot Status Changed: \(bootInfo.progressDescription)")
      lastBootInfo = bootInfo
      lastInfoUpdateDate = Date()
    }
  }

}

extension SimDeviceBootInfoStatus {

  /// The status as it reads in a log line.
  var statusName: String {
    switch self {
    case .booting:
      return "Booting"
    case .waitingOnBackboard:
      return "WaitingOnBackboard"
    case .waitingOnDataMigration:
      return "WaitingOnDataMigration"
    case .waitingOnSystemApp:
      return "WaitingOnSystemApp"
    case .finished:
      return "Finished"
    case .dataMigrationFailed:
      return "DataMigrationFailed"
    @unknown default:
      return "Unknown(\(rawValue))"
    }
  }
}

extension SimDeviceBootInfo {

  /// The boot's progress as it reads in a log line: the status and how long it has taken, plus the
  /// migration detail when that is what it is waiting on.
  var progressDescription: String {
    let progress = "\(status.statusName) | Elapsed \(bootElapsedTime)"
    guard status == .waitingOnDataMigration else {
      return progress
    }
    return "\(progress) | Migration Phase '\(migrationPhaseDescription ?? "")' | Migration Elapsed \(migrationElapsedTime)"
  }
}
